src/vdaf/poplar1.rs

```diff
@@ -204,47 +356,506 @@ impl<P: Prg<L>, const L: usize> Aggregator<L> for Poplar1<P, L> {
+        match out_share {
+            IdpfOutputShare::Inner(..) => panic!("tried to convert inner share into leaf field"),
+            IdpfOutputShare::Leaf(array) => array,
+        }
+    }
+}
+
+/// Construct a `Prng` with the given seed and info-string suffix.
+fn init_prng<I, B, F, P, const L: usize>(mut prg: P, info_suffix_parts: I) -> Prng<F, P::SeedStream>
+where
+    I: IntoIterator<Item = B>,
+    B: AsRef<[u8]>,
+    P: Prg<L>,
+    F: FieldElement,
+{
+    prg.update(VERSION); // info prefix
+    prg.update(&Poplar1::<P, L>::ID.to_be_bytes());
+    for info_suffix_part in info_suffix_parts.into_iter() {
+        prg.update(info_suffix_part.as_ref());
+    }
+    Prng::from_seed_stream(prg.into_seed_stream())
+}
+
+/// Derive shares of the correlated randomness for the next level of the IDPF tree.
+//
+// TODO(cjpatton) spec: Consider deriving the shares of a, b, c for each level directly from the
+// seed, rather than iteratively, as we do in Doplar. This would be more efficient for the
+// Aggregators. As long as the Client isn't significantly slower, this should be a win.
+fn compute_next_corr_shares<F: FieldElement + From<u64>, S: SeedStream>(
+    prng: &mut Prng<F, S>,
+    corr_prng_0: &mut Prng<F, S>,
+    corr_prng_1: &mut Prng<F, S>,
+    auth: F,
+) -> ([F; 2], [F; 2]) {
+    let two = F::from(2);
+    let a = corr_prng_0.get() + corr_prng_1.get();
+    let b = corr_prng_0.get() + corr_prng_1.get();
+    let c = corr_prng_0.get() + corr_prng_1.get();
+    let x = -two * a + auth; // A
+    let y = a * a + b - a * auth + c; // B
+    let corr_1 = [prng.get(), prng.get()];
+    let corr_0 = [x - corr_1[0], y - corr_1[1]];
+    (corr_0, corr_1)
+}
+
+/// Evaluate the IDPF at the given prefixes and compute the Aggregator's share of the sketch.
+fn eval_and_sketch<P, F, const L: usize>(
+    verify_key: &[u8; L],
+    agg_id: usize,
+    nonce: &[u8],
+    agg_param: &Poplar1AggregationParam,
+    public_share: &Poplar1PublicShare<L>,
+    idpf_key: &Seed<L>,
+    corr_prng: &mut Prng<F, P::SeedStream>,
+) -> Result<(Vec<F>, Vec<F>), VdafError>
+where
+    P: Prg<L>,
+    F: FieldElement,
+    [F; 2]: From<IdpfOutputShare<2, Field64, Field255>>,
+{
+    let nonce_len = nonce.len().try_into().map_err(|_| {
+        VdafError::Uncategorized(format!(
+            "nonce length cannot exceed 255 bytes ({})",
+            nonce.len()
+        ))
+    })?;
+
+    let level = u16::try_from(agg_param.level)
+        .map_err(|_| VdafError::Uncategorized(format!("level too deep ({})", agg_param.level)))?
+        .to_be_bytes();
+
+    // TODO(cjpatton) spec: Consider not encoding the prefixes here.
+    let mut verify_prng = init_prng::<_, _, _, _, L>(
+        P::init(verify_key),
+        [
+            DST_VERIFY_RANDOMNESS,
+            // TODO(cjpatton) spec: Drop length prefix if we decide to fix the nonce length.
+            &[nonce_len],
+            nonce,
+            &level,
+        ],
```

Comment:
Reviewer note: This value has changed to be more aligned with what I expect to land in VDAF-04. We decided yesterday that we don't need to implement VDAF-03 anyway.